unblob/handlers/archive/hp/bdl.py

```diff
@@ -0,0 +1,109 @@
+import io
+from pathlib import Path
+from typing import Iterable, Optional, Tuple
+
+from dissect.cstruct import Instance
+from structlog import get_logger
+
+from unblob.extractor import carve_chunk_to_file
+from unblob.file_utils import Endian, File, InvalidInputFormat, StructParser, snull
+from unblob.models import Chunk, Extractor, HexString, StructHandler, ValidChunk
+
+logger = get_logger()
+
+C_DEFINITIONS = r"""
+    typedef struct bdl_toc_entry {
+        uint64 offset;
+        uint64 size;
+    } bdl_toc_entry_t;
+
+    typedef struct bdl_header {
+        char magic[4];
+        uint16 major;
+        uint16 minor;
+        uint32 content_table_offset;
+        char unknown[4];
+        uint32 number_of_element;
+        uint32 unknowns_2[3];
+        char release[256];
+        char brand[256];
+        char device_id[256];
+        char unknown_3[9];
+        char version[256];
+        char revision[256];
+    } bdl_header_t;
+"""
+
+
+def is_valid_header(header: Instance) -> bool:
+    if header.major != 0x1:
+        return False
+    if header.minor != 0x1:
+        return False
+    try:
+        snull(header.brand).decode("utf-8")
+        snull(header.device_id).decode("utf-8")
+        snull(header.version).decode("utf-8")
+    except UnicodeDecodeError:
+        return False
+    return True
+
+
+class BDLExtractor(Extractor):
+    def __init__(self, header_struct: str):
+        self.header_struct = header_struct
+        self._struct_parser = StructParser(C_DEFINITIONS)
+
+    def extract(self, inpath: Path, outdir: Path):
+        with File.from_path(inpath) as file:
+            i = 0
+            for i, (offset, size) in enumerate(self.parse(file)):
+                file.seek(offset)
+                carve_chunk_to_file(
+                    file=file,
+                    chunk=Chunk(start_offset=offset, end_offset=offset + size),
+                    carve_path=outdir.joinpath(outdir.joinpath(Path(f"ipkg{i}"))),
+                )
+
+    def parse(self, file: File) -> Iterable[Tuple[int, int]]:
+        header = self._struct_parser.parse("bdl_header_t", file, Endian.LITTLE)
+        file.seek(header.content_table_offset, io.SEEK_SET)
+
+        entries = [
+            (entry.offset, entry.size)
+            for entry in (
+                self._struct_parser.parse("bdl_toc_entry_t", file, Endian.LITTLE)
+                for _ in range(header.number_of_element)
+            )
+        ]
```

Comment:
I feel like you could simplify this by doing everything in the `extract` function within a single loop.